[Merged by Bors] - feat(analysis/normed_space/spectrum and algebra/algebra/spectrum): prove properties of spectrum in a Banach algebra

j-loreaux opened the pull request:
Prove that the `spectrum` of an element in a Banach algebra is closed and bounded, hence compact when the scalar field is                               
proper. Compute the derivative of the `resolvent a` function in preparation for showing that the spectrum is nonempty when  the base field is ℂ. Define the `spectral_radius` and prove that it is bounded by the norm. Also rename the resolvent set to `resolvent_set` instead of `resolvent` so that it doesn't clash with the function name.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

j-loreaux committed:
feat(algebra/algebra/spectrum): lemmas when scalars are a field

Prove some properties of the spectrum when the underlying scalar ring
is a field, and mostly assuming the algebra is itself nontrivial.
Show that the spectrum of a scalar (i.e., `algebra_map 𝕜 A k`) is
the singleton `{k}`. Prove that `σ (a * b) \ {0} = σ (b * a) \ {0}`.

j-loreaux committed:
make changes suggested by review

j-loreaux committed:
rename `resolvent` to `resolvent_set`

This is for consistency with the naming in the
mathematical community because `resolvent` is
reserved for the function which acts on the
`resolvent_set` by taking each `r : R` to the
inverse of `algebra_map R A r - a`.

j-loreaux committed:
define `resolvent` and prove the basic equality

j-loreaux committed:
feat(analyis/normed_space/spectrum and algebra/algebra/spectrum):
prove properties of spectrum in a Banach algebra

Prove that the `spectrum` of an element in a Banach algebra
is closed and bounded, hence compact when the scalar field is
proper. Compute the derivative of the `resolvent a` function
in preparation for showing that the spectrum is compact when
the base field is ℂ. Define the `spectral_radius` and prove
that it is bounded by the norm.

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,182 @@
+  intros k' hk',
+  rw [metric.mem_ball, dist_eq_norm] at hk',
+  exact mem_resolvent_set_of_nearby hk hk',
+end
```
```suggestion
lemma is_open_resolvent_set (a : A) : is_open (ρ a) :=
units.is_open.preimage ((algebra_map_isometry 𝕜 A).continuous.sub continuous_const)
```

j-loreaux replied:
nice, so much simpler.

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,182 @@
+
+/-- The `resolvent_set` as a term of `opens 𝕜` -/
+def open_resolvent_set (a : A) : topological_space.opens 𝕜 :=
+⟨ρ a, is_open_resolvent_set a⟩
```
I think you can omit this one.  We usually use `is_open`; the bundled version `topological_space.opens` is only used in a few parts of the library (I think in sheaf theory, for example) where one needs a name for the whole collection.

j-loreaux replied:
okay, we probably will want the corresponding type for the spectrum though (nonempty compact) because then we can say that the spectrum is an upper semicontinuous function w.r.t. the Hausdorff metric. But we're not there yet.

hrmacbeth replied:
Not sure I follow -- for me, upper semicontinuity makes sense for functions into an ordered type, and "the type of compact subsets of C, equipped with the Hausdorff metric" does not have a natural order?  But indeed we're not there yet :)

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,182 @@
+⟨ρ a, is_open_resolvent_set a⟩
+
+lemma is_closed (a : A) : is_closed (σ a) :=
+is_open.is_closed_compl (is_open_resolvent_set a)
```
The so-called "projection notation" makes this kind of thing more efficient to express:
```suggestion
(is_open_resolvent_set a).is_closed_compl
```

j-loreaux replied:
done, I'm still learning exactly when using projection notation is available.

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,182 @@
+      ...            = ∥k∥⁻¹ * ∥a∥   : by rw normed_field.norm_inv
+      ...            < 1            : (inv_mul_lt_iff k_pos).mpr (by simp [h]),
+  have : is_unit (1 - ku⁻¹ • a), from (units.one_sub (ku⁻¹ • a) lt_one).is_unit,
+  rwa ←is_unit.smul_sub_iff_sub_inv_smul at this,
```
Maybe something like this?
```suggestion
  let ku := units.mk (↑ₐk) (↑ₐk⁻¹) sorry sorry,
  have hku : ∥-a∥ < ∥(↑ku⁻¹:A)∥⁻¹ := by simpa [ku, algebra_map_isometry] using h,
  simpa [ku, sub_eq_add_neg, algebra.algebra_map_eq_smul_one] using (ku.add (-a) hku).is_unit,
```
(`sorry`s because I don't know if there is a convenience constructor available to make `↑ₐk` a unit, maybe you've added it lately?)

j-loreaux replied:
yes, I found a nice thing: `units.map (↑ₐ).to_monoid_hom (units.mk0 k hk)` where `hk : k ≠ 0`

hrmacbeth replied:
Nice!

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,182 @@
+    `spectral_radius a = 0` by the definition of `Sup` on `ℝ`.-/
+noncomputable def spectral_radius (𝕜 : Type*) {A : Type*} [normed_field 𝕜] [ring A]
+  [algebra 𝕜 A] (a : A) : ℝ :=
+Sup (norm '' (spectrum 𝕜 a))
```
```suggestion
⨆ k ∈ spectrum 𝕜 a, ∥k∥
```
This means more or less the same thing, but the notation is nice, and there are some convenience lemmas available for it (keywords `supr` or perhaps `bsupr`)

j-loreaux replied:
changed, and defined it as an `ennreal` instead so that `supr` makes more sense.

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,182 @@
+begin
+  have ha' : (norm '' (σ a)).nonempty, from
+    ⟨∥ha.some∥,⟨ha.some,set.nonempty.some_mem ha,rfl⟩⟩,
+  have norm_up_bd : ∥a∥ ∈ upper_bounds (norm '' (σ a)), by
```
The style convention is to omit the `by`.

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,182 @@
+    { apply mem_upper_bounds.mpr,
+      rintros k ⟨k,⟨k_mem,k_eq⟩⟩,
+      subst k_eq,
+      exact norm_le_norm_of_mem k_mem },
```
```suggestion
  { rintros k ⟨k, ⟨k_mem, rfl⟩⟩,
    exact norm_le_norm_of_mem k_mem },
```
`rintros` can do more work for you here.  (Note, here and elsewhere, that the style convention is to use spaces in things separated by angle brackets.)

j-loreaux replied:
ah yes, someone mentioned this `rfl` syntax for `rcases` recently, I should have realized it applied to `rintros` also.

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,182 @@
+
+-- this lemma is actually true without the assumption `(σ a).nonempty`,
+-- but it would require a `by_cases` split, and it's only for the slightly
+-- silly reason that `Sup (∅ : set ℝ)` is defeq `0`.
```
The mathlib style is to include these silly cases in the lemma, too, so please do the case split.  (It just makes life easier for a user down the line, since they don't need to check side conditions.)

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,182 @@
+      ... ≤ C * ∥k'∥ * (c * C⁻¹) : mul_le_mul_of_nonneg_left (le_of_lt hk'.left) (by nlinarith [C_pos, norm_nonneg k'])
+      ... = (C * C⁻¹) * c * ∥k'∥ : by ring
+      ... = c * ∥k'∥ : by simp [mul_inv_cancel (ne_of_gt C_pos)],
+    },
```
And for this one, I think you missed the similar lemma `has_fderiv_at_ring_inverse` (it is certainly hard to find things in mathlib until you know the contents approximately already!)
```suggestion
  have H₁ : has_fderiv_at inverse _ (↑ₐk - a) := has_fderiv_at_ring_inverse hk.unit,
  have H₂ : has_deriv_at (λ k, ↑ₐk - a) 1 k,
  { simpa using (algebra.linear_map 𝕜 A).has_deriv_at.sub_const a },
  simpa [resolvent, hk.unit_spec, ← inverse_unit hk.unit] using H₁.comp_has_deriv_at k H₂,
```

j-loreaux replied:
ouch, yes, that would have made life easier.

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,182 @@
+
+open asymptotics normed_ring ring
+
+theorem resolvent_has_deriv_at {a : A} {k : 𝕜} (hk : k ∈ ρ a) :
```
```suggestion
theorem has_deriv_at_resolvent {a : A} {k : 𝕜} (hk : k ∈ ρ a) :
```
(this is the order our rather arbitrary naming convention has selected)

dupuisf commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,182 @@
+import algebra.algebra.spectrum
+import analysis.calculus.deriv
+/-!
+# The spectrum of elements in complete normed algebra
```
```suggestion
# The spectrum of elements in a complete normed algebra
```

dupuisf reviewed (commented):
Thanks! I don't really have much to add over Heather's comments, it otherwise looks good!

j-loreaux committed:
typo

j-loreaux committed:
redefine `spectral_radius` and improve theorem

also rewrote `spectral_radius_le_nnnorm` in order
to include the case when the spectrum is empty,
and simultaneously simplified the proof.

j-loreaux committed:
significantly simplify proof of `is_open (ρ a)`

j-loreaux committed:
use projection notation for `is_closed_compl`

j-loreaux committed:
simplify proof `mem_resolvent_of_norm_lt`

j-loreaux committed:
simplify and rename `has_deriv_at_resolvent`

j-loreaux committed:
update documentation

dupuisf commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,117 @@
+    coerced into an element of `ℝ≥0∞` so that it lives in a `complete_lattice`. Note that it
+    is possible for `spectrum 𝕜 a = ∅`. In this case, `spectral_radius a = 0`-/
+noncomputable def spectral_radius (𝕜 : Type*) {A : Type*} [normed_field 𝕜] [ring A]
+  [algebra 𝕜 A] (a : A) : ℝ≥0∞ :=
```
Wouldn't it make more sense for this to be a `nnreal` rather than a `ennreal`, since it's never infinite? (i.e. I don't see why it would be important for this to live in a complete lattice.)

j-loreaux replied:
A few reasons:

1. It gives you access to all the theorems about `supr` on complete lattices.
2. you can always reduce to `nnreal` if need be with `mod_cast` tactics since the definition casts it into an `ennreal` in first place (see, e.g., the proof of `spectral_radius_le_nnnorm`).
3. perhaps most importantly: unbounded operators can have unbounded spectrum, and at least that tau-measurable operators (in general unbounded) affiliated with a semifinite von Neumann algebra with faithful normal semifinite trace tau form a *-algebra. So really, I should have been allowing infinity in the first place anyway.

j-loreaux replied:
To really drive home the last point in a way which is probably more accessible: consider L^0([0,1]), the collection of measurable functions on the unit interval which are finite almost everywhere, with the usual quotient by equal almost everywhere equivalence. This is an algebra, and the spectrum of an element is precisely its essential range. So, if the element isn't L^infty, then its spectrum is unbounded.

hrmacbeth replied:
I find this last example really convincing -- and like Frederic I had been initially dubious when the rationale was expressed as "making it live in a complete lattice".  Let me suggest-change a docstring which emphasizes these points.

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,117 @@
+lemma mem_resolvent_of_norm_lt {a : A} {k : 𝕜} (h : ∥a∥ < ∥k∥) :
+  k ∈ ρ a :=
+begin
+  rw [resolvent_set,set.mem_set_of_eq,algebra.algebra_map_eq_smul_one],
```
```suggestion
  rw [resolvent_set, set.mem_set_of_eq, algebra.algebra_map_eq_smul_one],
```

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,117 @@
+
+lemma norm_le_norm_of_mem {a : A} {k : 𝕜} (hk : k ∈ σ a) :
+  ∥k∥ ≤ ∥a∥ :=
+le_of_not_lt (not_imp_not.mpr mem_resolvent_of_norm_lt hk)
```
tiny golf:
```suggestion
le_of_not_lt $ mt mem_resolvent_of_norm_lt hk
```

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,117 @@
+le_of_not_lt (not_imp_not.mpr mem_resolvent_of_norm_lt hk)
+
+lemma subset_closed_ball_norm (a : A) :
+  σ a ⊆ metric.closed_ball (0 : 𝕜) (∥a∥) :=
```
My own taste would be to incorporate this into the proof of the next lemma rather than having it as a stand-alone lemma.  But if you prefer as-is, don't worry about it.

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,117 @@
+theorem spectral_radius_le_nnnorm (a : A) :
+  spectral_radius 𝕜 a ≤ ∥a∥₊ :=
+begin
+  suffices h : ∀ (k : 𝕜) (hk : k ∈ σ a), (∥k∥₊ : ℝ≥0∞) ≤ ∥a∥₊,
```
```suggestion
  suffices h : ∀ k ∈ σ a, (∥k∥₊ : ℝ≥0∞) ≤ ∥a∥₊,
```

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,117 @@
+  suffices h : ∀ (k : 𝕜) (hk : k ∈ σ a), (∥k∥₊ : ℝ≥0∞) ≤ ∥a∥₊,
+  { exact bsupr_le h, },
+  { by_cases ha : (σ a).nonempty,
+    { intros _ hk, exact_mod_cast norm_le_norm_of_mem hk },
```
```suggestion
    { intros _ hk,
      exact_mod_cast norm_le_norm_of_mem hk },
```
Each tactic application goes on its own line.

hrmacbeth reviewed (commented):
Is this PR now actually independent of #10476?  If so, you could split out the lemmas from that PR and remove the dependency.

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,117 @@
+begin
+  rw [resolvent_set,set.mem_set_of_eq,algebra.algebra_map_eq_smul_one],
+  have hk : k ≠ 0 := ne_zero_of_norm_pos (by linarith [norm_nonneg a]),
+  let ku := units.map (↑ₐ).to_monoid_hom (units.mk0 k hk),
```
```suggestion
  let ku := (units.mk0 k hk).map (↑ₐ).to_monoid_hom,
```
more projection notation (untested)

j-loreaux replied:
doesn't seem to work

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,117 @@
+
+/-- The *spectral radius* is the supremum of the `nnnorm` (`∥⬝∥₊`) of elements in the spectrum,
+    coerced into an element of `ℝ≥0∞` so that it lives in a `complete_lattice`. Note that it
+    is possible for `spectrum 𝕜 a = ∅`. In this case, `spectral_radius a = 0`-/
```
```suggestion
/-- The *spectral radius* is the supremum of the `nnnorm` (`∥⬝∥₊`) of elements in the spectrum,
    coerced into an element of `ℝ≥0∞`. Note that it is possible for `spectrum 𝕜 a = ∅`. In this
    case, `spectral_radius a = 0`.  It is also possible that `spectrum 𝕜 a` be unbounded (though
    not for Banach algebras, see `spectrum.is_bounded`, below).  In this case,
    `spectral_radius a = ∞`. -/
```

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,117 @@
+
+## Main statements
+
+* `is_open_resolvent_set`: the resolvent set is open.
```
I believe that for all these definitions and statements in the module docstring, you should use namespacing, `spectrum.is_open_resolvent_set`, etc.  This ensures that when the documentation is auto-generated, it will include a link to the declaration.

hrmacbeth commented on src/analysis/normed_space/spectrum.lean:
```diff
@@ -0,0 +1,117 @@
+local notation `↑ₐ` := algebra_map 𝕜 A
+
+theorem has_deriv_at_resolvent {a : A} {k : 𝕜} (hk : k ∈ ρ a) :
+  has_deriv_at (resolvent a) (-(resolvent a k) * (resolvent a k)) k :=
```
Would it be conventional to express this as
```suggestion
  has_deriv_at (resolvent a) (-(resolvent a k) ^ 2) k :=
```
(Up to you.)

j-loreaux committed:
Apply suggestions from code review

improve documentation, mathlib style, minigolf

Co-authored-by: hrmacbeth <25316162+hrmacbeth@users.noreply.github.com>

j-loreaux committed:
namespace documentation for auto-linking

j-loreaux committed:
switch `mul_self` with `pow_two` in derivative

j-loreaux committed:
Revert "make changes suggested by review"

This reverts commit faff9b570a9bededb0c9d46118ef0481c9beb159.

j-loreaux committed:
Revert "feat(algebra/algebra/spectrum): lemmas when scalars are a field"

This reverts commit 580cc7c25bb488a2d38cebe91178c029c9189ce9.

j-loreaux committed:
fix lint error for line length
